logstash-core/src/main/java/org/logstash/plugins/inputs/Generator.java

```diff
@@ -0,0 +1,202 @@
+            PluginConfigSpec.arraySetting("lines");
+    public static final PluginConfigSpec<String> MESSAGE_CONFIG =
+            PluginConfigSpec.stringSetting("message", "Hello world!");
+    public static final PluginConfigSpec<Long> THREADS_CONFIG =
+            PluginConfigSpec.numSetting("threads", 1);
+    public static final PluginConfigSpec<Double> EPS_CONFIG =
+            PluginConfigSpec.floatSetting("eps", 0);
+
+    private final String hostname;
+    private final long count;
+    private final double eps;
+    private String id;
+    private long threads;
+    private volatile boolean stopRequested = false;
+    private CountDownLatch[] countDownLatches;
+    private String[] lines;
+    private int[] linesIndex;
+    private long[] sequence;
+    private ScheduledFuture<?>[] futures;
+    private List<Map<String, Object>> events;
+
+    /**
+     * Required constructor.
+     *
+     * @param id            Plugin id
+     * @param configuration Logstash Configuration
+     * @param context       Logstash Context
+     */
+    public Generator(final String id, final Configuration configuration, final Context context) {
+        this.id = id;
+        this.count = configuration.get(COUNT_CONFIG);
+        this.eps = configuration.get(EPS_CONFIG);
+        this.threads = configuration.get(THREADS_CONFIG);
+        if (this.threads < 1) {
+            throw new IllegalStateException("May not specify fewer than one generator thread");
+        }
+
+        String host;
+        try {
+            host = InetAddress.getLocalHost().getHostName();
+        } catch (UnknownHostException e) {
+            host = "[unknownHost]";
+        }
+        this.hostname = host;
+
+        // specifying "lines" will override "message"
+        List<Object> linesConfig = configuration.get(LINES_CONFIG);
+        if (linesConfig != null) {
+            lines = new String[linesConfig.size()];
+            for (int k = 0; k < linesConfig.size(); k++) {
+                lines[k] = (String) linesConfig.get(k);
+            }
+
+        } else {
+            lines = new String[]{configuration.get(MESSAGE_CONFIG)};
+        }
+    }
+
+    @Override
+    public void start(Consumer<Map<String, Object>> writer) {
+        if (eps > 0) {
+            startThrottledGenerator(writer);
+        } else {
+            startUnthrottledGenerator(writer);
+        }
+    }
+
+    private void startUnthrottledGenerator(Consumer<Map<String, Object>> writer) {
+        sequence = new long[(int) threads];
+        countDownLatches = new CountDownLatch[(int) threads];
+        events = new ArrayList<>();
+        linesIndex = new int[(int) threads];
+
+        for (int k = 0; k < threads; k++) {
+            Map<String, Object> event = new HashMap<>();
+            event.put("hostname", hostname);
+            event.put("thread_number", k);
+            events.add(event);
+            countDownLatches[k] = new CountDownLatch(1);
```

Comment:
Do we need to have a `CountdownLatch` per thread? Would it make more sense to have a single `CountDownLatch` that is initialised with the number of things to wait for, each of which counts down when it is ready to close down? That way we could await on a single instance during shutdown.